Move ElectrodesTable creation

stephprince opened the pull request:
Fix #155 by moving the creation of ElectrodesTable to a separate convenience function.

As part of these changes, I added some additional tests, a helper function for getting the size of a dataset, and a check for creating ElectricalSeries with electrodes indices outside of the possible range of the ElectrodesTable.

oruebel commented on docs/pages/userdocs/workflow.dox:
```diff
@@ -14,12 +14,14 @@
  * 3. Create the \ref AQNWB::NWB::NWBFile "NWBFile" object used for managing and creating NWB
  *    file contents.
- * 4. Create the \ref AQNWB::NWB::Container "Container" objects (e.g.,
+ * 4. Create the \ref AQNWB::NWB::ElectrodesTable "ElectrodesTable" object used for saving 
+ *    electrodes information. 
```
```suggestion
 * 4. Create the recording metadata stored in ``\general`` in NWB, e.g., the  
 *    \ref AQNWB::NWB::ElectrodesTable "ElectrodesTable" object used for 
 *    saving electrodes information. 
```

oruebel commented on docs/pages/userdocs/workflow.dox:
```diff
@@ -52,7 +54,22 @@
  *
  *
- * \subsection create_datasets 4. Create datasets and add to RecordingContainers.
+ * \subsection create_nwbfile 4. Create the ElectrodesTable object.
+ *
+ * Next, constructs the \ref AQNWB::NWB::ElectrodesTable "ElectrodesTable" object, 
+ * using the `recordingArrays` as an input. Note that in many cases, you will be recording from 
+ * all electrodes detected by the device and this `recordingArrays` structure will be the same as
+ * is used to create the \ref AQNWB::NWB::ElectricalSeries "ElectricalSeries". However in some cases
+ * (e.g. Neuropixels), you may be recording from a subset of all available electrodes. In this case,
+ * you should create the ref AQNWB::NWB::ElectrodesTable "ElectrodesTable" with all electrodes
+ * detected by the acquisition system, not just those being actively recorded from. This approach is
+ * critical for mapping information about which \ref AQNWB::NWB::ElectricalSeries "ElectricalSeries"
+ * were recorded from which electrodes.
```
```suggestion
 * \subsection create_recmeta 4. Create the recording metadata 
 * \subsubsection create_recmeta_ecephys Create the extracellular recording metadata
 * Next, constructs the \ref AQNWB::NWB::ElectrodesTable "ElectrodesTable" object, 
 * using the `recordingArrays` as an input. Note that in many cases, you will be recording from 
 * all electrodes detected by the device and this `recordingArrays` structure will be the same as
 * is used to create the \ref AQNWB::NWB::ElectricalSeries "ElectricalSeries". However in some cases
 * (e.g. Neuropixels or when using multiple probees), you may be recording from a subset of 
 * all available electrodes as part of a single \ref AQNWB::NWB::ElectricalSeries "ElectricalSeries". In this case,
 * you should create the ref AQNWB::NWB::ElectrodesTable "ElectrodesTable" with all electrodes
 * detected by the acquisition system, not just those being actively recorded from. This approach is
 * critical for mapping information about which \ref AQNWB::NWB::ElectricalSeries "ElectricalSeries"
 * were recorded from which electrodes.
```

oruebel commented on src/io/BaseIO.hpp:
```diff
@@ -441,6 +441,13 @@ class BaseIO
   virtual std::unique_ptr<BaseRecordingData> getDataSet(
       const std::string& path) = 0;
 
+  /**
+   * @brief Returns the size of the dataset for each dimension.
+   * @param path The location in the file of the dataset.
+   * @return The dataset size.
+   */
+  virtual std::vector<SizeType> getDatasetSize(const std::string path) = 0;
```
- It would be nice to have this function support both datasets and attributes. In that case, I would rename it to `getStorageObjectShape`.  
- It would be nice to also add the following function to `ReadDataWrapper` to make this new function convenient to use on read

```  
inline std::vector<SizeType> getShape() const
{
   return m_io->getShape(m_path);
}

- We should be consistent in how we use `shape` and `size`. I think we should use `shape` to refer to the sizes along all dimensions and `size` to refer to the total number of elements.

stephprince replied:
> It would be nice to have this function support both datasets and attributes. In that case, I would rename it to getStorageObjectShape.
It would be nice to also add the following function to ReadDataWrapper to make this new function convenient to use on read

That makes sense, I will add functionality for attributes and use `shape` to refer to sizes along all dimensions

oruebel commented on src/nwb/NWBFile.cpp:
```diff
@@ -28,7 +28,7 @@ constexpr SizeType CHUNK_XSIZE =
     8;  // TODO - replace with io settings input
 
-std::vector<SizeType> NWBFile::emptyContainerIndexes = {};
+std::vector<SizeType> NWBFile::m_emptyContainerIndexes = {};
```
Maybe `Types.h` is a better place for this or alternatively, I think we can also just explicitly use "std::vector<SizeType>()" as default argument (at least it seems that `m_emptyContainerIndexes` is only used to set default arguments of functions, i.e., we could just write:

```
Status createAnnotationSeries(
      std::vector<std::string> recordingNames,
      RecordingContainers* recordingContainers = nullptr,
      std::vector<SizeType>& containerIndexes = std::vector<SizeType>());
```

stephprince replied:
If I remember correctly, I think I set this up this way because you cannot pass a temporary empty initializer list as a default argument to a non constant reference parameter. Creating the empty vector in advance was a solution for that. Another potential option is to change the parameter to an rvalue reference, but then I think there might be additional complications in terms of transferring ownership.

stephprince commented:
Based on our discussion, I made `containerIndexes` a required argument and removed the static `m_emptyContainerIndexes` variable. I think `recordingContainers` should likely also be required as an input for the user so I removed the default value for that parameter as well.

These changes will now also fix #162, if we want to take a different approach or discuss that particular issue more I can move the changes to `getStorageObjectsShape` to a separate PR.

stephprince commented:
@oruebel this is ready for you to re-review

oruebel commented on docs/pages/userdocs/workflow.dox:
```diff
@@ -14,12 +14,15 @@
  * 3. Create the \ref AQNWB::NWB::NWBFile "NWBFile" object used for managing and creating NWB
  *    file contents.
- * 4. Create the \ref AQNWB::NWB::Container "Container" objects (e.g.,
+ * 4. Create the recording metadata stored in ``\general`` in NWB, e.g., the  
```
```suggestion
 * 4. Create the recording metadata stored in ``/general`` in NWB, e.g., the  
```

oruebel commented on src/io/BaseIO.hpp:
```diff
@@ -441,6 +441,14 @@ class BaseIO
       const std::string& path) = 0;
 
+  /**
+   * @brief Returns the size of the dataset for each dimension.
+   * @param path The location in the file of the dataset.
+   * @return The dataset shape.
```
```suggestion
   * @brief Returns the size of the dataset or attribute for each dimension.
   * @param path The location of the dataset or attribute in the file
   * @return The shape of the dataset or attribute.
```

oruebel commented on src/io/hdf5/HDF5IO.hpp:
```diff
@@ -268,6 +268,13 @@ class HDF5IO : public BaseIO
   std::unique_ptr<IO::BaseRecordingData> getDataSet(
       const std::string& path) override;
 
+  /**
+   * @brief Returns the size of the dataset for each dimension.
+   * @param path The location in the file of the dataset.
+   * @return The dataset shape.
+   */
```
```suggestion
   * @brief Returns the size of the dataset or attribute for each dimension.
   * @param path The location of the dataset or attribute in the file
   * @return The shape of the dataset or attribute.
```

oruebel commented on src/nwb/NWBFile.cpp:
```diff
@@ -172,6 +170,43 @@ Status NWBFile::createFileStructure(const std::string& identifierText,
+  }
+
+  // write electrodes information to datasets
+  // (requires that ElectrodeGroup data is written)
```
If the ElectrodeGroup data must be written, should this function then call ``m_io->flush()`` here first?

stephprince replied:
```suggestion
  // (requires that ElectrodeGroup data is initialized)
```

oruebel commented:
I added a couple of minor comments (mostly updates to docs), but otherwise this looks good. 

Note, there will be conflicts with #163

stephprince commented:
@oruebel I responded to the last comment, I think this is ready to be merged once you approve
